pkg/hubble/relay/pool/manager.go

```diff
@@ -186,6 +194,41 @@ func (m *PeerManager) manageConnections() {
 	}
 }
 
+func (m *PeerManager) reportConnectionStatus() {
+	connTimer, connTimerDone := inctimer.New()
+	defer connTimerDone()
+	for {
+		select {
+		case <-m.stop:
+			return
+		case <-connTimer.After(m.opts.connStatusInterval):
+			m.mu.RLock()
+			connStates := make(map[connectivity.State]uint32)
+			var nilConnPeersNum uint32 = 0
+			for _, p := range m.peers {
+				p.mu.Lock()
+				if p.conn == nil {
+					nilConnPeersNum++
+					p.mu.Unlock()
+					continue
+				}
+				state := p.conn.GetState()
+				// if does not exists, num is 0 (zero value)
+				// which happens to be a perfect initial value
+				num, _ := connStates[state]
+				num++
+				connStates[state] = num
+				p.mu.Unlock()
+			}
+			m.mu.RUnlock()
+
+			m.mu.Lock()
```

Comment:
nit: this lock is unnecessary, as the `ObservePeerConnectionStatus` only does the `Set` method on `prometheus GaugeVec`, and prometheus client is already thread-safe.

Reply:
The lock was definitely needed, because `ObservePeerConnectionStatus` was not thread-safe - even though setting Prometheus gauge for a particular label was thread-safe, this function was calling `Set` multiple times and so having states {"IDLE": 3, "READY": 1} and {"IDLE": 1, "READY": 3} being observed at the same time, you could end up with states {"IDLE": 3, "READY": 3} being displayed through gauge metric.

Now though I just made the `ObservePeerConnectionStatus` thread-safe itself, so that the lock used in metrics won't block other PeerManager operations. Useful nit to that, so thanks.

Reply:
I like your solution with having a separate mutex for the metric gauge setter. Thanks.